File: src/main/java/io/github/xf8b/xf8bot/listeners/MessageListener.java
/*
 * Copyright (c) 2020 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.listeners;

import com.google.common.util.concurrent.ThreadFactoryBuilder;
import com.mongodb.MongoCommandException;
import com.mongodb.client.model.Filters;
import com.mongodb.reactivestreams.client.MongoCollection;
import discord4j.core.event.domain.message.MessageCreateEvent;
import discord4j.core.object.entity.Guild;
import discord4j.core.object.entity.Member;
import discord4j.core.object.entity.Message;
import discord4j.rest.http.client.ClientException;
import io.github.xf8b.xf8bot.Xf8bot;
import io.github.xf8b.xf8bot.api.commands.AbstractCommand;
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent;
import io.github.xf8b.xf8bot.api.commands.CommandRegistry;
import io.github.xf8b.xf8bot.api.commands.arguments.Argument;
import io.github.xf8b.xf8bot.api.commands.flags.Flag;
import io.github.xf8b.xf8bot.commands.InfoCommand;
import io.github.xf8b.xf8bot.exceptions.ThisShouldNotHaveBeenThrownException;
import io.github.xf8b.xf8bot.settings.CommandHandlerChecks;
import io.github.xf8b.xf8bot.settings.DisableChecks;
import io.github.xf8b.xf8bot.util.PermissionUtil;
import io.github.xf8b.xf8bot.util.Result;
import io.github.xf8b.xf8bot.util.parser.ArgumentParser;
import io.github.xf8b.xf8bot.util.parser.FlagParser;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.bson.Document;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Slf4j
@RequiredArgsConstructor
public class MessageListener {
    @NotNull
    private final Xf8bot xf8bot;
    @NotNull
    private final CommandRegistry commandRegistry;
    private static final ExecutorService COMMAND_THREAD_POOL = Executors.newCachedThreadPool(new ThreadFactoryBuilder()
            .setNameFormat("Command Pool Thread-%d")
            .build());
    private static final ArgumentParser ARGUMENT_PARSER = new ArgumentParser();
    private static final FlagParser FLAG_PARSER = new FlagParser();

    public static void shutdownCommandThreadPool() {
        COMMAND_THREAD_POOL.shutdown();
    }

    public Mono<MessageCreateEvent> onMessageCreateEvent(@NotNull MessageCreateEvent event) {
        //TODO: reactify all the classes
        //TODO: make exception handler
        //TODO: add spam protection
        Message message = event.getMessage();
        String content = message.getContent();
        Guild guild = event.getGuild().block();
        String guildId = guild.getId().asString();
        if (content.trim().equals("<@!" + event.getClient().getSelfId().asString() + "> help")) {
            InfoCommand commandHandler = commandRegistry.getCommandHandler(InfoCommand.class);
            return onCommandFired(event, commandHandler, guildId, content).thenReturn(event);
        }
        MongoCollection<Document> mongoCollection = xf8bot.getMongoDatabase().getCollection("prefixes");
        Mono.from(mongoCollection.find(Filters.eq("guildId", Long.parseLong(guildId))))
                .cast(Object.class)
                .switchIfEmpty(Mono.from(mongoCollection.insertOne(new Document()
                        .append("guildId", Long.parseLong(guildId))
                        .append("prefix", Xf8bot.DEFAULT_PREFIX))))
                .block();
        String commandType = content.trim().split(" ")[0];
        for (AbstractCommand commandHandler : commandRegistry) {
            String name = commandHandler.getNameWithPrefix(xf8bot, guildId);
            List<String> aliases = commandHandler.getAliasesWithPrefixes(xf8bot, guildId);
            if (commandType.equalsIgnoreCase(name)) {
                return onCommandFired(event, commandHandler, guildId, content).thenReturn(event);
            } else if (!aliases.isEmpty()) {
                for (String alias : aliases) {
                    if (commandType.equalsIgnoreCase(alias)) {
                        return onCommandFired(event, commandHandler, guildId, content).thenReturn(event);
                    }
                }
            }
        }
        return Mono.just(event);
    }

    private Mono<Void> onCommandFired(@NotNull MessageCreateEvent event, @NotNull AbstractCommand commandHandler, @NotNull String guildId, @NotNull String content) {
        Result<Map<Flag<?>, Object>> flagParseResult = FLAG_PARSER.parse(commandHandler, content);
        Result<Map<Argument<?>, Object>> argumentParseResult = ARGUMENT_PARSER.parse(commandHandler, content);
        if (flagParseResult.getResultType() != Result.ResultType.FAILURE && argumentParseResult.getResultType() != Result.ResultType.FAILURE) {
            CommandFiredEvent commandFiredEvent = new CommandFiredEvent(
                    xf8bot,
                    flagParseResult.getResult(),
                    argumentParseResult.getResult(),
                    event
            );
            return commandFiredEvent.getGuild()
                    .flatMap(Guild::getSelfMember)
                    .flatMap(Member::getBasePermissions)
                    .map(permissions -> permissions.containsAll(commandHandler.getBotRequiredPermissions()))
                    .flatMap(bool -> {
                        if (commandHandler.getClass().getAnnotation(DisableChecks.class) != null) {
                            if (Arrays.asList(commandHandler.getClass().getAnnotation(DisableChecks.class).value())
                                    .contains(CommandHandlerChecks.BOT_HAS_REQUIRED_PERMISSIONS)) {
                                return Mono.just(bool);
                            }
                        }
                        if (!bool) {
                            commandFiredEvent.getChannel()
                                    .flatMap(messageChannel -> messageChannel
                                            .createMessage(String.format("Could not execute command \"%s\" because of insufficient permissions!", commandHandler.getNameWithPrefix(xf8bot, guildId))))
                                    .subscribe();
                            return Mono.empty();
                        } else {
                            return Mono.just(true);
                        }
                    })
                    .flatMap(bool -> {
                        if (commandHandler.getClass().getAnnotation(DisableChecks.class) != null) {
                            if (Arrays.asList(commandHandler.getClass().getAnnotation(DisableChecks.class).value())
                                    .contains(CommandHandlerChecks.IS_ADMINISTRATOR)) {
                                return Mono.just(bool);
                            }
                        }
                        if (commandHandler.requiresAdministrator()) {
                            if (!PermissionUtil.canMemberUseCommand(xf8bot, commandFiredEvent.getGuild().block(), commandFiredEvent.getMember().get(), commandHandler)) {
                                commandFiredEvent.getChannel()
                                        .flatMap(messageChannel -> messageChannel.createMessage("Sorry, you don't have high enough permissions.")).subscribe();
                                return Mono.empty();
                            }
                        }
                        return Mono.just(bool);
                    })
                    .flatMap(bool -> {
                        if (commandHandler.getClass().getAnnotation(DisableChecks.class) != null) {
                            if (Arrays.asList(commandHandler.getClass().getAnnotation(DisableChecks.class).value())
                                    .contains(CommandHandlerChecks.IS_BOT_ADMINISTRATOR)) {
                                return Mono.just(bool);
                            }
                        }
                        if (commandHandler.isBotAdministratorOnly()) {
                            if (!commandFiredEvent.getXf8bot().isBotAdministrator(commandFiredEvent.getMember().get().getId())) {
                                commandFiredEvent.getChannel()
                                        .flatMap(messageChannel -> messageChannel.createMessage("Sorry, you aren't a administrator of xf8bot."))
                                        .subscribe();
                                return Mono.empty();
                            }
                        }
                        return Mono.just(bool);
                    })
                    .flatMap(bool -> {
                        if (commandHandler.getClass().getAnnotation(DisableChecks.class) != null) {
                            if (Arrays.asList(commandHandler.getClass().getAnnotation(DisableChecks.class).value())
                                    .contains(CommandHandlerChecks.SURPASSES_MINIMUM_AMOUNT_OF_ARGUMENTS)) {
                                return Mono.just(bool);
                            }
                        }
                        if (content.trim().split(" ").length < commandHandler.getMinimumAmountOfArgs() + 1) {
                            commandFiredEvent.getMessage().getChannel()
                                    .flatMap(messageChannel -> messageChannel.createMessage("Huh? Could you repeat that? The usage of this command is: `" + commandHandler.getUsageWithPrefix(xf8bot, guildId) + "`."))
                                    .subscribe();
                            return Mono.empty();
                        } else {
                            return Mono.just(bool);
                        }
                    })
                    .flatMap($ -> commandHandler.onCommandFired(commandFiredEvent))
                    .doOnError(throwable -> LOGGER.error("An error happened while handling commands!", throwable))
                    .onErrorResume(ClientException.class, t -> commandFiredEvent.getChannel()
                            .flatMap(messageChannel -> messageChannel.createMessage("Client exception happened while handling command: " + t.getErrorResponse().get().getFields()))
                            .then())
                    .onErrorResume(MongoCommandException.class, t -> commandFiredEvent.getChannel()
                            .flatMap(messageChannel -> messageChannel.createMessage("Database error happened while handling command: " + t.getErrorCodeName()))
                            .then())
                    .onErrorResume(ThisShouldNotHaveBeenThrownException.class, t -> commandFiredEvent.getChannel()
                            .flatMap(messageChannel -> messageChannel.createMessage("Something has horribly gone wrong. Please report this to the bot developer with the log."))
                            .then())
                    .onErrorResume(t -> commandFiredEvent.getChannel()
                            .flatMap(messageChannel -> messageChannel.createMessage("Exception happened while handling command: " + t.getMessage()))
                            .then());
        } else {
            if (flagParseResult.getResultType() == Result.ResultType.FAILURE) {
                return event.getMessage()
                        .getChannel()
                        .flatMap(channel -> channel.createMessage(flagParseResult.getErrorMessage()))
                        .then();
            } else if (argumentParseResult.getResultType() == Result.ResultType.FAILURE) {
                return event.getMessage()
                        .getChannel()
                        .flatMap(channel -> channel.createMessage(argumentParseResult.getErrorMessage()))
                        .then();
            } else {
                throw new ThisShouldNotHaveBeenThrownException();
            }
        }
    }
}

File: src/main/kotlin/io/github/xf8b/xf8bot/listeners/MessageListener.kt
/*
 * Copyright (c) 2020, 2021 xf8b.
 *
 * This file is part of xf8bot.
 *
 * xf8bot is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * xf8bot is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with xf8bot.  If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.xf8b.xf8bot.listeners

import discord4j.core.`object`.entity.User
import discord4j.core.`object`.entity.channel.MessageChannel
import discord4j.core.event.ReactiveEventAdapter
import discord4j.core.event.domain.message.MessageCreateEvent
import discord4j.rest.http.client.ClientException
import io.github.xf8b.utils.exceptions.UnexpectedException
import io.github.xf8b.xf8bot.Xf8bot
import io.github.xf8b.xf8bot.api.commands.Command
import io.github.xf8b.xf8bot.api.commands.CommandFiredEvent
import io.github.xf8b.xf8bot.api.commands.CommandRegistry
import io.github.xf8b.xf8bot.api.commands.parsers.ArgumentCommandInputParser
import io.github.xf8b.xf8bot.api.commands.parsers.FlagCommandInputParser
import io.github.xf8b.xf8bot.commands.info.InfoCommand
import io.github.xf8b.xf8bot.database.actions.find.FindDisabledCommandAction
import io.github.xf8b.xf8bot.util.*
import io.github.xf8b.xf8bot.util.extensions.getAdministratorLevel
import io.github.xf8b.xf8bot.util.extensions.toSnowflake
import io.github.xf8b.xf8bot.util.extensions.updatedRows
import reactor.core.publisher.Mono
import reactor.kotlin.core.publisher.onErrorResume
import reactor.kotlin.core.publisher.toFlux

class MessageListener(
    private val xf8bot: Xf8bot,
    private val commandRegistry: CommandRegistry
) : ReactiveEventAdapter() {
    override fun onMessageCreate(event: MessageCreateEvent): Mono<Void> {
        if (event.message.content.isBlank()
            || event.member.isEmpty
            || event.message.author.map(User::isBot).orElse(true)
        ) {
            return Mono.empty()
        }

        if (event.message.content matches "<@!?${event.client.selfId.asString()}> help".toRegex()) {
            val infoCommand = commandRegistry.findRegisteredWithType<InfoCommand>()

            return onCommandFired(event, infoCommand, "")
        }

        val message = event.message
        val guildId = event.guildId.get().asString()
        val contentMono = xf8bot.prefixCache
            .get(guildId.toSnowflake())
            .map { prefix ->
                if (message.content.startsWith(prefix)) message.content.removePrefix(prefix).trim()
                else ""
            }

        // FIXME: fix levels

        return /*handleLevels(event).thenEmpty(*/ contentMono.flatMap { content ->
            content.findCommand().flatMap { command ->
                xf8bot.botDatabase.execute(FindDisabledCommandAction(guildId.toSnowflake(), command))
                    .singleOrEmpty()
                    .filterWhen { result -> result.updatedRows }
                    .flatMap {
                        event.member.get()
                            .getAdministratorLevel(xf8bot)
                            .map { administratorLevel -> administratorLevel >= 4 }
                    }
                    .defaultIfEmpty(true)
                    .flatMap { allowedToRun ->
                        if (allowedToRun) {
                            event.message.channel
                                .flatMap(MessageChannel::type)
                                .thenEmpty(onCommandFired(event, command, content))
                        } else {
                            event.message.channel
                                .flatMap { it.createMessage("Sorry, but this command has been disabled by an administrator.") }
                                .then()
                        }
                    }
            }
        }
        // )
    }

    private fun String.findCommand(): Mono<Command> = commandRegistry.toFlux()
        .filter { command ->
            command.rawName.equals(this.split(" ")[0], ignoreCase = true)
                    || command.rawAliases.any { alias -> alias.equals(this.split(" ")[0], ignoreCase = true) }
        }
        .singleOrEmpty()

    /*
    private fun handleLevels(event: MessageCreateEvent): Mono<Void> = xf8bot.botDatabase
        .execute(GetXpAction(event.guildId.get(), event.member.get().id))
        .filter { it.isNotEmpty() }
        .flatMapMany { it.toFlux() }
        .flatMap { it.map { row, _ -> row } }
        .map { it["xp", java.lang.Long::class.java] }
        .cast<Long>()
        .switchIfEmpty(
            xf8bot.botDatabase
                .execute(AddXpAction(guildId = event.guildId.get(), memberId = event.member.get().id))
                .cast()
        )
        .flatMap { previousXp ->
            val previousLevel = LevelsCalculator.xpToLevels(previousXp as Long)
            val newXp = previousXp + LevelsCalculator.randomXp(5, 25)
            val newLevel = LevelsCalculator.xpToLevels(newXp)

            xf8bot.botDatabase
                .execute(UpdateXpAction(guildId = event.guildId.get(), memberId = event.member.get().id, xp = newXp))
                .then(Mono.zip(previousLevel.toMono(), newLevel.toMono()).flatMap { levelTuple ->
                    if (levelTuple.t1 < levelTuple.t2) {
                        event.message.channel.flatMap {
                            it.createMessage("Congratulations, you've leveled up from ${levelTuple.t1} to ${levelTuple.t2}!")
                        }
                    } else {
                        Mono.empty()
                    }
                })
        }
        .then()
     */

    private fun onCommandFired(event: MessageCreateEvent, command: Command, content: String): Mono<Void> {
        val cleanedContent = content.split(" ").drop(1).joinToString(separator = " ")

        val flagParseResult = FLAG_PARSER.parse(command, cleanedContent)
        val argumentParseResult = ARGUMENT_PARSER.parse(command, cleanedContent)

        return if (flagParseResult.isSuccess() && argumentParseResult.isSuccess()) {
            val commandFiredEvent = CommandFiredEvent(
                event,
                xf8bot,
                flagParseResult.result!!,
                argumentParseResult.result!!
            )

            commandFiredEvent.guild
                .filterWhen { Checks.doesBotHavePermissionsRequired(command, it.selfMember, commandFiredEvent.channel) }
                .filterWhen { Checks.doesMemberHaveCorrectAdministratorLevel(command, commandFiredEvent) }
                .filterWhen {
                    Checks.canMemberUseBotAdministratorOnlyCommand(
                        command,
                        xf8bot,
                        event.member.get(),
                        event.message.channel
                    )
                }
                .filterWhen { Checks.isThereEnoughArguments(command, commandFiredEvent) }
                .flatMap { command.onCommandFired(commandFiredEvent) }
                .doOnError { LOGGER.error("An error happened while handling command ${command.rawName}!", it) }
                .onErrorResume(ClientException::class) { exception ->
                    commandFiredEvent.channel
                        .flatMap { it.createMessage("Client exception happened while handling command: ${exception.status}: ${exception.errorResponse.get().fields}") }
                        .then()
                }
                .onErrorResume(UnexpectedException::class) {
                    commandFiredEvent.channel
                        .flatMap { it.createMessage("Something has horribly gone wrong. Please report this to the bot developer with the log.") }
                        .then()
                }
                .onErrorResume { throwable ->
                    commandFiredEvent.channel
                        .flatMap { it.createMessage("Exception happened while handling command: ${throwable.message}") }
                        .then()
                }
        } else {
            when {
                flagParseResult.isFailure() -> event.message.channel
                    .flatMap { it.createMessage(flagParseResult.errorMessage) }
                    .then()

                argumentParseResult.isFailure() -> event.message.channel
                    .flatMap { it.createMessage(argumentParseResult.errorMessage) }
                    .then()

                else -> throw UnexpectedException()
            }
        }
    }

    companion object {
        private val ARGUMENT_PARSER = ArgumentCommandInputParser()
        private val FLAG_PARSER = FlagCommandInputParser()
        private val LOGGER by LoggerDelegate()
    }
}
